protocol/src/traits/executor.rs

```diff
@@ -22,12 +22,39 @@ pub struct ExecutorParams {
     pub cycles_limit: u64,
 }
 
-#[derive(Debug, Clone)]
-pub struct ExecResp {
-    pub ret:      String,
-    pub is_error: bool,
+#[derive(Debug, Clone, Default)]
+pub struct ServiceResponse<T: Default> {
```

Comment:
Use enum?

```
pub enum ServiceResponse {
  Success(T),
  Error({ code: u32, err: String})
}

impl ServiceResponse {
  fn is_err() -> bool
}
```

Reply:
adjust ServiceResponse to : 
```rust

#[derive(Debug, Clone, Default)]
pub struct ServiceResponse<T: Default> {
    pub code:          u64,
    pub succeed_data:  T,
    pub error_message: String,
}

impl<T: Default> ServiceResponse<T> {
    pub fn from_error(code: u64, error_message: String) -> Self {
        Self {
            code,
            succeed_data: T::default(),
            error_message,
        }
    }

    pub fn from_succeed(succeed_data: T) -> Self {
        Self {
            code: 0,
            succeed_data,
            error_message: "".to_owned(),
        }
    }
}
```

Please check again

Reply:
I think `succeed` and `error_message` are somehow duplicated.
If we want Backward Compatibility, we can still use `ret` and `is_error`, or we can use `code` and `message`, when it is 0, the message is succeed_data, or it is error_message.

Reply:
> Use enum?
> 
> ```
> pub enum ServiceResponse {
>   Success{ data: T},
>   Error({ code: u32, err: String})
> }
> 
> impl ServiceResponse {
>   fn is_err() -> bool
> }
> ```

The rust style enum is not common in other languages, I think we should avoid using it in interface layer, or it will introduce many troubles in serde/deserde and other sdk.

Reply:
> The rust style enum is not common in other languages, I think we should avoid using it in interface layer, or it will introduce many troubles in serde/deserde and other sdk.

So why don't you use `c`?

Reply:
> I think `succeed_data` and `error_message` are somehow duplicated.
> If we want Backward Compatibility, we can still use `ret` and `is_error`, or we can use `code` and `message`, when it is 0, the message is succeed_data, otherwise it is error_message.

Because of `code`, i think we can't make backward compatibility.